### PR TITLE
Share to Social Media

### LINKED ISSUE #94: [backend] post on twitter/gist

### PR DESCRIPTION
**Pull request by kittybest**
- change `oauth()` function in `ceremony context` (Q: should we put the posting stuff in this context? or new a context related to `oauth` stuff)
- add posting function in backend in `github` router
- adjust the oauth-token scope for github
-

close #94  #95 .

### CODE REVIEW

**Comment by kittybest** [comment]
I already deployed the backend to `https://http.ceremony.unirep.io` and I think it works well now.
@vivianjeng

**Comment by ChialiT** [comment]
image


this does posted on twitter, but the behavior is weird tho.
first it asked me to authorized, and the project is called UniRep Social (we should be called UniRep)
then it didn't let me preview the tweet, it just post it without my consensus. 
the OG image didn't show either, I think the link is wrong too.

should be using [Anon] - if user uses no auth method.

**Comment by kittybest** [comment]
@ChialiT 
- Should be [Anon] --> this part is an issue I opened: #103 , is a problem not only related to this `share-sns` branch, so I don't want to fix it this  PR.
- Unirep Social is because I use my old twitter oauth token... I cannot apply for a second token since I don't pay for it, but of course I can change the name to **Unirep Trusted Setup** since unirep social is not working now.
- without preview --> I cannot imagine how will it work with preview? The user will be redirected to twitter and they can go to anywhere else instead of posting and might lost their way back to the ceremony? Do you have example website of how this work?

**Comment by ChialiT** [comment]
> @ChialiT
> 
> * Should be [Anon] --> this part is an issue I opened: [[Frontend] Username #103](https://github.com/Unirep/trusted-setup/issues/103) , is a problem not only related to this `share-sns` branch, so I don't want to fix it this  PR.
> * Unirep Social is because I use my old twitter oauth token... I cannot apply for a second token since I don't pay for it, but of course I can change the name to **Unirep Trusted Setup** since unirep social is not working now.
> * without preview --> I cannot imagine how will it work with preview? The user will be redirected to twitter and they can go to anywhere else instead of posting and might lost their way back to the ceremony? Do you have example website of how this work?

image
the preview just an typical OG image, if you have set up some kinda SEO on the site, it should be render.

like the kzg ceremony
image

**Comment by kittybest** [comment]
@ChialiT 
Oh I don't know what is an OG picture and now I know. I will add it. (I thought it will appear directly after we setup our frontend settings... but it seems like we could do something to make it more customizable.)
Btw, I think we could do like **we post for the user on gist/twitter ➡️ back to ceremony webpage but display a popup box providing a button to go to their post**, now the users need to check by themselves and it's not that convenient.
Also I just think of that the displayed URL in the post is wrong (that's the url of the backend not frontend), I will modify it now.

**Comment by CJ-Rose** [comment]
my twitter post looks like this
Screenshot 2023-10-03 at 12 49 35 PM

if i saw someone post this i would not have any idea what it was about. i think we should change the copy here, it would be more clear if we simply say 'I just contributed to UniRep trusted setup ceremony.' and then 'Generate your (own) verse and help secure the UniRep protocol here:'

and the second line of text should not be repeated both in the body of the post and in the link.

**Comment by ChialiT** [comment]
> my twitter post looks like this Screenshot 2023-10-03 at 12 49 35 PM
> 
> if i saw someone post this i would not have any idea what it was about. i think we should change the copy here, it would be more clear if we simply say 'I just contributed to UniRep trusted setup ceremony.' and then 'Generate your (own) verse and help secure the UniRep protocol here:'
> 
> and the second line of text should not be repeated both in the body of the post and in the link.

yes, more clear copy. thank you!

**Comment by vivianjeng** [comment]
截圖 2023-10-04 上午9 16 21

when I click **Github** it redirects to `localhost:8000`

**Comment by ChialiT** [comment]
✨ UniRep Ceremony ✨
I just contributed to UniRep trusted setup ceremony.
Contribute to help secure the UniRep protocol here: https://ceremony.unirep.io

- meta description: Generate your own verse and help secure the UniRep protocol.

**Comment by vivianjeng** [comment]
I think it can have a logo,  description and url to unirep page (or PSE page) in the Twitter Oauth token

image

**Comment by kittybest** [comment]
@ChialiT what kind of logo do you think we should have for our app? I could upload to twitter/discord/github oauth app setting.
Also need the website description for `og:description` and oath app description.

**Comment by kittybest** [comment]
> Generate your own verse and help secure the UniRep protocol.

fixed.

**Comment by ChialiT** [comment]
> @ChialiT what kind of logo do you think we should have for our app? I could upload to twitter/discord/github oauth app setting. Also need the website description for `og:description` and oath app description.

description: Generate your own verse and help secure the UniRep protocol.

and lets use the same favicon as the app, u should be able to generate the png from figma.

**Comment by kittybest** [comment]
> > @ChialiT what kind of logo do you think we should have for our app? I could upload to twitter/discord/github oauth app setting. Also need the website description for `og:description` and oath app description.
> 
> description: Generate your own verse and help secure the UniRep protocol.
> 
> and lets use the same favicon as the app, u should be able to generate the png from figma.

Logo and description updated.

**Comment by CJ-Rose** [comment]
Screenshot 2023-10-05 at 9 19 03 AM

still no image on my post, does it appear for others?

we should remove the second line of text: "Generate your own verse and help secure the UniRep protocol"
--and put that text inside the link with the image instead of what is there now: "Generate your verse.....to knitting together the constellation...."   <-- this is the part that doesn't make sense

File: packages/frontend/src/pages/Contribute.jsx [hunk 1]
```diff
@@ -264,15 +304,63 @@ export default observer(() => {
...
+  }
+
   return (
     <>
       <ToastContainer position="top-center" theme="colored" />
+      <Popup
+        open={postMessage.platform}
+        onClose={() => setPostMessage({})}
+        title="Post Successfully!"
+        content={`Post on ${postMessage.platform} successfully.`}
```
Comment (CJ-Rose) [inline packages/frontend/src/pages/Contribute.jsx]: change to:

title="Post Successful!"
content={'Your contribution was posted on ${postMessage.platform}. '}

File: packages/frontend/src/pages/Contribute.jsx [hunk 2]
```diff
@@ -264,15 +304,63 @@ export default observer(() => {
...
+  const postOnTwitter = async () => {
+    const url = new URL('/oauth/twitter', HTTP_SERVER)
+    url.searchParams.set('token', ceremony.authToken)
+    const currentUrl = new URL(window.location.href)
+    const dest = new URL('/contribute', currentUrl.origin)
+    url.searchParams.set('redirectDestination', dest.toString())
+    url.searchParams.set(
+      'content',
+      `✨ UniRep Ceremony ✨\nI just contributed to UniRep trusted setup ceremony.\nContribute to help secure the UniRep protocol here: ${currentUrl.origin}\n\nGenerate your own verse and help secure the UniRep protocol.`
+    )
```
Comment (CJ-Rose) [inline packages/frontend/src/pages/Contribute.jsx]: remove the last line here:
'Generate your own verse and help secure the UniRep protocol'

--and move that text to replace what is currently attached to the link in `index.html`

File: packages/frontend/public/index.html [hunk 3]
```diff
@@ -12,6 +12,15 @@
...
     />
     <title>trusted setup</title>
+    <meta property="og:title" content="Unirep Trusted Setup" />
+    <meta property="og:type" content="website" />
+    <meta property="og:url" content="https://ceremony.unirep.io" />
+    <meta property="og:image" content="og.png" />
+    <meta property="og:site_name" content="Unirep Ceremony" />
+    <meta
+      property="og:description"
+      content="Generate your verse to the UniRep Ceremony to knitting together the constellation that unveil our journey."
```
Comment (CJ-Rose) [inline packages/frontend/public/index.html]: replace `content` with:

'Generate your own verse and help secure the UniRep protocol.'

**Review by CJ-Rose — CHANGES_REQUESTED** [review]

**Review by vivianjeng — COMMENTED** [review]
The webpack will not include the `og.png` in the output folder

it should work like this
https://1bbd14f7.trusted-setup-og.pages.dev/

and you should be able to preview it in this link
https://www.opengraph.xyz/url/https%3A%2F%2F1bbd14f7.trusted-setup-og.pages.dev%2F

**Pull request merged**